ably/subscription.go

```diff
@@ -0,0 +1,280 @@
+	sub.queue = append(sub.queue, msg)
+	if sleeping {
+		sub.sleep <- struct{}{}
+	}
+}
+
+func (sub *Subscription) pop() (msg interface{}, n int) {
+	sub.mtx.Lock()
+	defer sub.mtx.Unlock()
+	if n = len(sub.queue); n == 0 {
+		return nil, 0
+	}
+	msg, sub.queue = sub.queue[0], sub.queue[1:]
+	return msg, n
+}
+
+func (sub *Subscription) loop() {
+	channel := reflect.ValueOf(sub.channel)
+	for range sub.sleep {
+		for msg, n := sub.pop(); n != 0; msg, n = sub.pop() {
+			channel.Send(reflect.ValueOf(msg))
+		}
+	}
+}
+
+var (
+	subsAll     struct{}
+	subsAllKeys = []interface{}{subsAll}
+)
+
+func namesToKeys(names []string) []interface{} {
+	if len(names) == 0 {
+		return nil
+	}
+	keys := make([]interface{}, 0, len(names))
+	for _, name := range names {
+		// Ignore empty names.
+		if name != "" {
+			keys = append(keys, name)
+		}
+	}
+	return keys
+}
+
+func statesToKeys(states []proto.PresenceState) []interface{} {
+	if len(states) == 0 {
+		return nil
+	}
+	keys := make([]interface{}, 0, len(states))
+	for _, state := range states {
+		keys = append(keys, state)
+	}
+	return keys
+}
+
+type subscriptions struct {
+	typ reflect.Type
+	mtx sync.Mutex
+	all map[interface{}]map[*Subscription]struct{}
+}
+
+func newSubscriptions(typ reflect.Type) *subscriptions {
+	return &subscriptions{
+		typ: typ,
+		all: make(map[interface{}]map[*Subscription]struct{}),
+	}
+}
+
+func (subs *subscriptions) close() {
+	for _, subs := range subs.all {
+		for sub := range subs {
+			// Stop is idempotent, no need to keep track which sub was already
+			// stopped.
+			sub.close(false)
+		}
+	}
+	// Unsubscribe all channels by creating new sub map for future use.
+	subs.all = make(map[interface{}]map[*Subscription]struct{})
+}
+
+func (subs *subscriptions) subscribe(keys ...interface{}) (*Subscription, error) {
+	unsubscribe := func(sub *Subscription) { subs.unsubscribe(false, sub, keys...) }
+	sub := newSubscription(subs.typ, unsubscribe)
+	if len(keys) == 0 {
+		keys = subsAllKeys
+	}
+	subs.mtx.Lock()
+	for _, key := range keys {
+		all, ok := subs.all[key]
+		if !ok {
+			all = make(map[*Subscription]struct{})
+			subs.all[key] = all
+		}
+		all[sub] = struct{}{}
+	}
+	subs.mtx.Unlock()
+	return sub, nil
+}
+
+func (subs *subscriptions) unsubscribe(stop bool, sub *Subscription, keys ...interface{}) {
+	if sub.typ != subs.typ {
+		panic(errInvalidType{typ: sub.typ})
+	}
+	if len(keys) == 0 {
+		keys = subsAllKeys
+	}
+	subs.mtx.Lock()
+	for _, key := range keys {
+		delete(subs.all[key], sub)
+		if len(subs.all[key]) == 0 {
+			delete(subs.all, key)
+		}
+	}
+	// Don't try to stop when we got here from the (*Subscription).Close method.
+	if stop {
+		count := 0
+		for _, all := range subs.all {
+			if _, ok := all[sub]; ok {
+				count++
+			}
+		}
+		// Stop subscription if it no longer listens for messages.
+		if count == 0 {
+			sub.close(false)
+		}
+	}
+	subs.mtx.Unlock()
+}
+
+func (subs *subscriptions) messageEnqueue(msg *proto.ProtocolMessage) {
```

Comment:
Unfortunately it’s not possible. You would need to abstract away iteration over items and key retrieval from a single item - it would be more code (one interface + two implementations), it would create intermediate structure for the sake of interface ([]interface{}) and it would be definitely more complex to understand than just two functions with two for loops.